cql3/functions/aggregate_fcts.cc

```diff
@@ -463,6 +464,45 @@ aggregate_fcts::make_min_function(data_type io_type) {
     );
 }
 
+function_name
+aggregate_fcts::first_function_name() {
+    return function_name::native_function("$$first$$");
+}
+
+shared_ptr<aggregate_function>
+aggregate_fcts::make_first_function(data_type io_type) {
+    io_type = io_type->without_reversed().shared_from_this();
+    // The function's state is a one-element tuple containing the value, if the tuple
+    // itself is null then the an input hasn't been seen yet
+    auto state_type = data_type(tuple_type_impl::get_instance({io_type}));
```

Comment:
Yes. We need to distinguish between "we saw a first value, and it was NULL" and "we did not see a first value". The comment was supposed to explain it, but maybe it's lacking.

Reply:
Ok, I did see the comment but wasn't sure if you used a tuple just because of this reason, or there is another purpose for this tuple - perhaps to keep multiple values. I guess there isn't.

I still think the more "proper" way to do this first() aggregator was to have an accumulator being a tuple (key, value), where the null key (not null tuple) means nothing yet, and the reducer also works. But I guess it's not important enough to make this first() perfect.

Reply:
It's possible to do that indeed. But I don't think it's worth the effort, it will practically never happen.